Userland/Utilities/run-tests.cpp

```diff
@@ -0,0 +1,333 @@
+        auto result = m_exclude_regex.match(file_path, PosixFlags::Global);
+        if (!result.success) // must NOT match the regex to be a valid test file
+            paths.append(file_path);
+    });
+    quick_sort(paths);
+    return paths;
+}
+
+bool TestRunner::should_skip_test(const LexicalPath& test_path)
+{
+    for (const String& dir : m_skip_directories) {
+        if (test_path.dirname().contains(dir))
+            return true;
+    }
+    for (const String& file : m_skip_files) {
+        if (test_path.basename().contains(file))
+            return true;
+    }
+    return false;
+}
+
+void TestRunner::do_run_single_test(const String& test_path)
+{
+    g_currently_running_test = test_path;
+
+    auto test_result = run_test_file(test_path);
+
+    switch (test_result.result) {
+    case Test::Result::Pass:
+        ++m_counts.tests_passed;
+        break;
+    case Test::Result::Skip:
+        ++m_counts.tests_skipped;
+        break;
+    case Test::Result::Fail:
+        ++m_counts.tests_failed;
+        break;
+    case Test::Result::Crashed:
+        ++m_counts.tests_failed; // FIXME: tests_crashed
+        break;
+    }
+    if (test_result.result != Test::Result::Skip)
+        ++m_counts.files_total;
+
+    m_total_elapsed_time_in_ms += test_result.time_taken;
+
+    bool crashed_or_failed = test_result.result == Test::Result::Fail || test_result.result == Test::Result::Crashed;
+    bool print_stdout_stderr = crashed_or_failed || m_print_all_output;
+    if (crashed_or_failed) {
+        print_modifiers({ BG_RED, FG_BLACK, FG_BOLD });
+        out("{}", test_result.result == Test::Result::Fail ? " FAIL  " : "CRASHED");
+        print_modifiers({ CLEAR });
+    } else {
+        print_modifiers({ BG_GREEN, FG_BLACK, FG_BOLD });
+        out(" PASS ");
+        print_modifiers({ CLEAR });
+    }
+
+    out(" {}", LexicalPath::relative_path(test_path, m_test_root));
+
+    print_modifiers({ CLEAR, ITALIC, FG_GRAY });
+    if (test_result.time_taken < 1000) {
+        outln(" ({}ms)", static_cast<int>(test_result.time_taken));
+    } else {
+        outln(" ({:3}s)", test_result.time_taken / 1000.0);
+    }
+    print_modifiers({ CLEAR });
+
+    if (test_result.result != Test::Result::Pass) {
+        print_modifiers({ FG_GRAY, FG_BOLD });
+        out("         Test:   ");
+        if (crashed_or_failed) {
+            print_modifiers({ CLEAR, FG_RED });
+            outln("{} ({})", test_result.file_path.basename(), test_result.result == Test::Result::Fail ? "failed" : "crashed");
+        } else {
+            print_modifiers({ CLEAR, FG_ORANGE });
+            outln("{} (skipped)", test_result.file_path.basename());
+        }
+        print_modifiers({ CLEAR });
+    }
+
+    // Make sure our clear modifiers goes through before we dump file output via write(2)
+    fflush(stdout);
+
+    if (print_stdout_stderr && test_result.stdout_err_fd > 0) {
+        int ret = lseek(test_result.stdout_err_fd, 0, SEEK_SET);
```

Comment:
I dunno if it's actually expected or not but I *have* to lseek here. Otherwise we just get insta-EOF and write nothing. I wouldn't expect that the fd that was written to in the child would have its seek position at the end in the parent? Haven't tested in linux though 🤷‍♂️